pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -759,16 +765,141 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
 	return false, nil
 }
 
-func setSoMark(fd int, secId identity.NumericIdentity) error {
-	mark := linux_defaults.MagicMarkIdentity
+func setSoMarks(fd int, ipFamily ipfamily.IPFamily, secId identity.NumericIdentity) error {
+	if err := unix.SetsockoptInt(fd, ipFamily.SocketOptsFamily, ipFamily.SocketOptsTransparent, 1); err != nil {
+		return fmt.Errorf("setsockopt(IP_TRANSPARENT) for %s failed: %w", ipFamily.Name, err)
+	}
+
+	mark := linux_defaults.MagicMarkEgress
 	mark |= int(uint32(secId&0xFFFF)<<16 | uint32((secId&0xFF0000)>>16))
 	err := unix.SetsockoptInt(fd, unix.SOL_SOCKET, unix.SO_MARK, mark)
 	if err != nil {
 		return fmt.Errorf("error setting SO_MARK: %w", err)
 	}
+
+	if err := unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_REUSEADDR, 1); err != nil {
+		return fmt.Errorf("setsockopt(SO_REUSEADDR) failed: %w", err)
+	}
+	if !option.Config.EnableBPFTProxy {
+		if err := unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_REUSEPORT, 1); err != nil {
+			return fmt.Errorf("setsockopt(SO_REUSEPORT) failed: %w", err)
+		}
+	}
+
 	return nil
 }
 
+var errNoReader = errors.New("Reader stopped")
+
+type Response struct {
+	*dns.Msg
+	err error
+}
+
+// A Client keeps state for concurrent transactions on the same upstream connection.
+type Client struct {
+	dns.Client
+	refcount int // protected by proxy's lock
```

Comment:
A wider lock is needed to manage the map of the (shared) clients anyway. Logically that lock is on the same scope as the proxy's lock, but maybe I should introduce another lock in `DNSProxy` for this purpose?

Reply:
I agree that the lock is necessary for the client cache. I worry that this implies serialising all DNS requests, when they previously were not.

Specifically, it looks to me that we'll be holding the `DNSProxy` lock while creating the socket (syscall, AFAIK), which sounds like it'll regress performance. Do you have a way to compare performance before and after this change?

Reply:
I do not have a way to compare performance before and after this change, unfortunately.

I'll see if I can minimize these critical sections in any way. It looks like on close the lock must be held while closing to avoid the case where a new connection with the same 5-tuple is created before the old one is actually closed.

Reply:
Can we have a map of locks for the current client keys? This way the proxy-wide lock only protects adding the lock to map (on first DNS request with given key), client creation can be done while holding the client-key-specific lock, which should help with parallelizing DNS client creation.

Reply:
Moved Dial to happen under client's own lock instead of the proxy lock, so all related system call overhead is not gone from the proxy's critical section. Unfortunately it looks like the related close must happen under the lock governing the map of clients, otherwise a bind error is possible.

Reply:
Now that the syscalls are out of the `NewDNSClient()` this is essentially the same as having a map of locks for the currently in-use keys, IMO.